File: library/src/main/java/com/getstream/sdk/chat/utils/ChannelListDiffCallback.java
package com.getstream.sdk.chat.utils;

import android.util.Log;

import androidx.recyclerview.widget.DiffUtil;

import com.getstream.sdk.chat.model.Channel;

import java.util.List;

public class ChannelListDiffCallback extends DiffUtil.Callback {
    private final static String TAG = ChannelListDiffCallback.class.getSimpleName();
    private List<Channel> oldList, newList;

    public ChannelListDiffCallback(List<Channel> oldList, List<Channel> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList == null ? 0 : oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList == null ? 0 : newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).equals(newList.get(newItemPosition));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        // read state
        // last message
        // channel members
        // channel name
        // channel extra data
        // TODO: fix this
        Channel oldChannel = oldList.get(oldItemPosition);
        Channel newChannel =  newList.get(newItemPosition);

//        Log.i(TAG, "areContentsTheSame?");
//        Log.i(TAG, oldChannel.getLastMessageDate().toString());
//        Log.i(TAG, newChannel.getLastMessageDate().toString());

        if (!oldChannel.getLastMessageDate().equals(newChannel.getLastMessageDate())) {
            Log.i(TAG, "LastMessage is different");
            return false;
        }

        if (!oldChannel.getName().contentEquals(newChannel.getName())) {
            Log.i(TAG, "channel name is different");
            return false;
        }

        return false;
    }
}

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/adapter/diff/ChannelListDiffCallback.kt
package io.getstream.chat.android.ui.channel.list.adapter.diff

import androidx.recyclerview.widget.DiffUtil
import io.getstream.chat.android.client.models.Channel
import io.getstream.chat.android.client.models.name
import io.getstream.chat.android.ui.utils.extensions.getLastMessage
import io.getstream.chat.android.ui.utils.extensions.getLastMessageReadCount
import io.getstream.chat.android.ui.utils.extensions.getUsers
public class ChannelListDiffCallback(
    private val oldList: List<Channel>,
    private val newList: List<Channel>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].cid == newList[newItemPosition].cid
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldChannel = oldList[oldItemPosition]
        val newChannel = newList[newItemPosition]

        return listOf(
            cidEqual(oldChannel, newChannel),
            nameEqual(oldChannel, newChannel),
            userEqual(oldChannel, newChannel),
            updateEqual(oldChannel, newChannel),
            extraDataEqual(oldChannel, newChannel),
            lastMessageEqual(oldChannel, newChannel),
            lastReadEqual(oldChannel, newChannel),
        ).all { it }
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val oldChannel = oldList[oldItemPosition]
        val newChannel = newList[newItemPosition]

        return ChannelItemDiff(
            nameChanged = !nameEqual(oldChannel, newChannel),
            avatarViewChanged = !userEqual(oldChannel, newChannel),
            readStateChanged = !lastReadEqual(oldChannel, newChannel),
            lastMessageChanged = !lastMessageEqual(oldChannel, newChannel)
        )
    }

    private fun cidEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.cid == newChannel.cid

    private fun nameEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.name == newChannel.name

    private fun userEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.getUsers() == newChannel.getUsers()

    private fun updateEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.updatedAt?.equals(newChannel.updatedAt) == true

    private fun extraDataEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.extraData == newChannel.extraData

    private fun lastMessageEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.getLastMessage() == newChannel.getLastMessage()

    private fun lastReadEqual(oldChannel: Channel, newChannel: Channel) =
        oldChannel.getLastMessageReadCount() == newChannel.getLastMessageReadCount()
}
